relay-quotas/src/rate_limiter.rs

```diff
@@ -0,0 +1,571 @@
+    max_limit: Option<u64>,
+}
+
+impl RateLimiter {
+    /// Creates a new `RateLimiter` instance.
+    pub fn new(pool: RedisPool) -> Self {
+        RateLimiter {
+            pool,
+            script: Arc::new(load_lua_script()),
+            max_limit: None,
+        }
+    }
+
+    /// Sets the maximum rate limit in seconds.
+    ///
+    /// By default, this rate limiter will return rate limits based on the quotas' `window` fields.
+    /// If a maximum rate limit is set, this limit is bounded.
+    pub fn max_limit(mut self, max_limit: Option<u64>) -> Self {
+        self.max_limit = max_limit;
+        self
+    }
+
+    /// Checks whether any of the quotas in effect for the given project and project key has been
+    /// exceeded and records consumption of the quota.
+    ///
+    /// By invoking this method, the caller signals that data is being ingested and needs to be
+    /// counted against the quota. This increment happens atomically if none of the quotas have been
+    /// exceeded. Otherwise, a rate limit is returned and data is not counted against the quotas.
+    ///
+    /// If no key is specified, then only organization-wide and project-wide quotas are checked. If
+    /// a key is specified, then key-quotas are also checked.
+    pub fn is_rate_limited(
+        &self,
+        quotas: &[Quota],
+        scoping: &ItemScoping,
+    ) -> Result<RateLimits, RateLimitingError> {
+        let timestamp = UnixTimestamp::now();
+
+        let mut invocation = self.script.prepare_invoke();
+        let mut tracked_quotas = Vec::new();
+        let mut rate_limits = RateLimits::new();
+
+        for quota in quotas {
+            if !quota.matches(scoping) {
+                // Silently skip all quotas that do not apply to this item.
+            } else if quota.limit == Some(0) {
+                // A zero-sized quota is strongest. Do not call into Redis at all, and do not
+                // increment any keys, as one quota has reached capacity (this is how regular quotas
+                // behave as well).
+                let retry_after = self.retry_after(REJECT_ALL_SECS);
+                rate_limits.add(RateLimit::from_quota(quota, scoping, retry_after));
+            } else if let Some(quota) = RedisQuota::new(quota, scoping, timestamp) {
+                // Remaining quotas are expected to be trackable in Redis.
+                let key = quota.key();
+                let refund_key = get_refunded_quota_key(&key);
+
+                invocation.key(key);
+                invocation.arg(quota.limit());
+
+                invocation.key(refund_key);
+                invocation.arg(quota.expiry().as_secs());
+
+                tracked_quotas.push(quota);
+            } else {
+                // This quota is neither a static reject-all, nor can it be tracked in Redis due to
+                // missing fields. We're skipping this for forward-compatibility.
+                log::warn!("skipping unsupported quota");
```

Comment:
can you push some extra data into sentry-sdk scope or put them into the log msg? Depending on what would group better